client/lib/create-selector/index.js

```diff
@@ -88,28 +86,25 @@ export default function createSelector(
 	getDependants = DEFAULT_GET_DEPENDANTS,
 	getCacheKey = DEFAULT_GET_CACHE_KEY
 ) {
-	const memoizedSelector = memoize( selector, getCacheKey );
-	let lastDependants;
+	const dependentsPerKey = new Map();
+	const memo = new Map();
 
 	if ( Array.isArray( getDependants ) ) {
 		getDependants = makeSelectorFromArray( getDependants );
 	}
 
-	return Object.assign(
-		function( state, ...args ) {
-			let currentDependants = getDependants( state, ...args );
-			if ( ! Array.isArray( currentDependants ) ) {
-				currentDependants = [ currentDependants ];
-			}
-
-			if ( lastDependants && ! shallowEqual( currentDependants, lastDependants ) ) {
-				memoizedSelector.cache.clear();
-			}
+	const memoizedSelector = function( state, ...args ) {
+		const cacheKey = getCacheKey( state, ...args );
+		const currentDependants = [].concat( getDependants( state, ...args ) );
+		const prevDependents = dependentsPerKey.get( cacheKey );
 
-			lastDependants = currentDependants;
+		if ( ! memo.has( cacheKey ) || ! shallowEqual( currentDependants, prevDependents ) ) {
+			memo.set( cacheKey, selector( state, ...args ) );
+			dependentsPerKey.set( cacheKey, currentDependants );
```

Comment:
Could the `memo` and `dependentsPerKey` maps be merged into one? We always update both of them at the same time, and also lookup in them at the same time. Let's do the map operations at once.

A crazy idea: instead of the data, we could store selector functions into the map!
```js
function createKeyedSelector( selector, getDependants, getKey ) {
	const selectorMap = new Map();
	return ( state, ...args ) => {
		const key = getKey( args );
		let keySelector = selectorMap.get( key );
		if ( ! keySelector ) {
			keySelector = createSelector(
				state => selector( state, ...args ),
				state => getDependants( state, ...args )
			);
			selectorMap.set( key, keySelector );
		}
		return keySelector( state );
	}
}
```
This should be equivalent to your implementation, although it consumes more memory. A keyed selector is just a map of non-keyed selectors. These sub-selectors remember the values of `prevDependants` and `memoizedValue`.

Reply:
I actually had an implementation earlier that I had scratched where I combined the maps into one!

They both get keyed at the same time by the same key, so i created literal objects and destructured them at the get/set points like:
```js
memo.set( cacheKey, { dependents: getDependents(...), value: selector(...) } )
```

Does this selectorMap version of the code solve either of the two main issues?
1. potentially calling getDependents twice 
2. inhibiting gc by storing outdated references to state inside of a map

Reply:
> Does this selectorMap version of the code solve either of the two main issues?

Unfortunately, it doesn't. It only combines two maps into one and shows the interesting, but otherwise not very useful fact that a keyed selector can be thought of as a map of non-keyed selectors.